Petsc interface

cmacmackin opened the pull request:
This is the PETSc interface which I implemented as part of #1803, which @ZedThree requested I split into a separate PR. It addresses issue #1771 and allows vectors and arrays to be set using the local BOUT++ index objects for that processor. The wrapper will automatically handle conversions to global indexing for PETSc. It also handles the interpolation weights needed when calculating along-field values in the y-direction. It has been extensively unit tested. The unit tests for the index conversion ability required some modest refactoring of the Mesh class, making some methods virtual which weren't previously and adding methods which wrap some MPI calls (so that they can be overridden for testing purposes). This allows the parallel behaviour of the indexing to be tested while running the unit tests in serial.

There were a number of specific comments Peter made about my code on the old PR. I will address those and then push the changes to this branch.

The basic call-sequence of my PETSc interface can be seen below:
![PetscInterfaceCall](https://user-images.githubusercontent.com/8145213/66742385-68e30600-ee6f-11e9-9ac9-122509328dcd.png)
Basically, I need the indexing operator to return something to which matrix/vector values can be assigned. When assigning to that object, it would then call the PETSc setter routines. While something like this could have been accomplished with a single method on PetscMatrix along the lines of 
```C++
  setElement(ind_type& index1, ind_type& index2, BoutReal val);
```
 the approach I took allowed something which feels more natural and is thus more readable.

Peter had said
> Also, historically `Mesh` has been a bit of dumping ground for new behaviour, and we'd quite like to avoid adding new things to it if at all possible. Do you think there are alternative ways of achieving the same functionality?

It should be possible to create a a new class which provides access to the MPI methods (or add them to BoutComm). The reason I took the approach I did was because it for minimal changes to existing calls to MPI routines within the code by ensuring the wrapper methods were in the same class/namespace as the Mesh routines which call them. I don't have any objections to placing the wrappers on a separate class, i just didn't feel confident enough to make such a disruptive change on my own.

cmacmackin commented:
Most of the specific issues raised on PR #1803 have now been resolved on this branch. The main ones that remain are

- [ ] adding an `operator BoutReal()` method to `PetscMatrix::Element`, if possible
- [ ] placing MPI calls onto their own class (e.g., `BoutComm`), rather than `BoutMesh`
- [ ] placing the protected `BoutMesh::BoutMesh()` constructor elsewhere?
- [ ] writing unit tests for `operator BoutReal()` on the `Element` classes
- [ ] deciding how to handle assignment of Element objects to each other
- [ ] anything else which gets flagged up during review

ZedThree commented:
Currently failing to compile because it turns out `std::less<void>` in `FakeParallelMesh` is C++14 only, and we currently only support C++11 (this will be changing after the next release!).
Luckily, I don't think `std::less<void>` is required, so you should just be able to delete it.

I pass `CXXFLAGS=-std=c++14` to `./configure` to force C++11 compatibility.

---

There's a strange double-free bug coming from the unit tests. I'm not sure I completely understand, but it looks like it's something to do with the way `createFakeProcessors` is setting `bout::globals::mesh`. I fixed it by adding a destructor to the test that resets the global mesh to `nullptr`:

```cpp
  virtual ~ParallelIndexerTest() {
    bout::globals::mesh = nullptr;
  }
```

Google recommend adding a virtual destructor to all test fixtures, btw. It's never been a problem for me, but it's probably wise.

---

The `ParallelIndexerTest` is quite noisy due to some of the functions its calling. There's a helper class, `WithQuietOutput` which can be used to silence particular output streams for the duration of a test. Just add some member variables to the fixture:
```cpp
WithQuietOutput info{output_info}, warn{output_warn}, progress{output_progress}, all{output};
```

`TestMixedSetting` and `TestDestroy` for the vector and matrix tests is also noisy, but this is from PETSc. If you set `PetscErrorPrintf = PetscErrorPrintfNone` in the fixture constructor, and `PetscErrorPrintf = PetscErrorPrintfDefault` in the destructor, that should silence them.
I'm guessing these are supposed to raise PETSc errors?

cmacmackin commented:
> Currently failing to compile because it turns out `std::less<void>` in `FakeParallelMesh` is C++14 only, and we currently only support C++11 (this will be changing after the next release!).
> Luckily, I don't think `std::less<void>` is required, so you should just be able to delete it.
> 
> I pass `CXXFLAGS=-std=c++14` to `./configure` to force C++11 compatibility.

Yes, I noticed this problem on the CI server. Good to see that it's an easy fix.

> There's a strange double-free bug coming from the unit tests. I'm not sure I completely understand, but it looks like it's something to do with the way `createFakeProcessors` is setting `bout::globals::mesh`. I fixed it by adding a destructor to the test that resets the global mesh to `nullptr`:
> 
> ```c++
>   virtual ~ParallelIndexerTest() {
>     bout::globals::mesh = nullptr;
>   }
> ```
> 
> Google recommend adding a virtual destructor to all test fixtures, btw. It's never been a problem for me, but it's probably wise.

Where do you see this bug? I'm not getting it when I run on my system. Also, do you have a reference for the Google recommendation? I'd be interested to read why.

> The `ParallelIndexerTest` is quite noisy due to some of the functions its calling. There's a helper class, `WithQuietOutput` which can be used to silence particular output streams for the duration of a test. Just add some member variables to the fixture:
> 
> ```c++
> WithQuietOutput info{output_info}, warn{output_warn}, progress{output_progress}, all{output};
> ```
> 
> `TestMixedSetting` and `TestDestroy` for the vector and matrix tests is also noisy, but this is from PETSc. If you set `PetscErrorPrintf = PetscErrorPrintfNone` in the fixture constructor, and `PetscErrorPrintf = PetscErrorPrintfDefault` in the destructor, that should silence them.
> I'm guessing these are supposed to raise PETSc errors?

Thanks for mentioning these. The noise was bothering me quite a lot, especially where it consists of PETSc error messages (which, as you say, are supposed to be raised, as I'm checking that improper use results in an exception being thrown). I had tried to figure out how to suppress the noise from PETSc but didn't find anything when I Googled, so not sure where `PetscErrorPrintf` is hiding in the documentation.

cmacmackin commented:
> Currently failing to compile because it turns out `std::less<void>` in `FakeParallelMesh` is C++14 only, and we currently only support C++11 (this will be changing after the next release!).
> Luckily, I don't think `std::less<void>` is required, so you should just be able to delete it.
> 
> I pass `CXXFLAGS=-std=c++14` to `./configure` to force C++11 compatibility.

I saw this in the CI logs. Glad to know it's an easy fix

> There's a strange double-free bug coming from the unit tests. I'm not sure I completely understand, but it looks like it's something to do with the way `createFakeProcessors` is setting `bout::globals::mesh`. I fixed it by adding a destructor to the test that resets the global mesh to `nullptr`:
> 
> ```c++
>   virtual ~ParallelIndexerTest() {
>     bout::globals::mesh = nullptr;
>   }
> ```
> 
> Google recommend adding a virtual destructor to all test fixtures, btw. It's never been a problem for me, but it's probably wise.

I'm not seeing this bug when I run on my machine. Might be something to do with the compiler version? Also, do you have a reference for that Google advice? I'd be interested to read the reasoning. 

> The `ParallelIndexerTest` is quite noisy due to some of the functions its calling. There's a helper class, `WithQuietOutput` which can be used to silence particular output streams for the duration of a test. Just add some member variables to the fixture:
> 
> ```c++
> WithQuietOutput info{output_info}, warn{output_warn}, progress{output_progress}, all{output};
> ```
> 
> `TestMixedSetting` and `TestDestroy` for the vector and matrix tests is also noisy, but this is from PETSc. If you set `PetscErrorPrintf = PetscErrorPrintfNone` in the fixture constructor, and `PetscErrorPrintf = PetscErrorPrintfDefault` in the destructor, that should silence them.
> I'm guessing these are supposed to raise PETSc errors?

I'd been rather irritated by the noise. Good to know there are ways to silence it. Yes, those PETSc errors are supposed to happen: I'm checking that an exception is thrown when the routines are used inappropriately. I'd Googled to try to find a way to suppress those warnings but hadn't found anything, so those variables must be somewhat hidden in the documentation.

ZedThree commented:
> Where do you see this bug? I'm not getting it when I run on my system.

It was a bit tricky to reproduce, as it involved multiple tests interacting so it could be due to tests being run in a different order. Perhaps try running `./serial_tests --gtest_shuffle --gtest_repeat=100` to see if that uncovers it? You could also use valgrind, but it doesn't look like the unit tests are even close to being valgrind clean at the minute. You would also need a suppression file to quiet the warnings from MPI as well.

> Also, do you have a reference for the Google recommendation? I'd be interested to read why.

I can't seem to find an explicit recommendation specifically for googletest now, so I'm probably misremembering, but the gist of it is that base classes should have virtual destructors to ensure they get called by derived classes. Each `TEST_F` in googletest gets turned into a subclass of the fixture. If the destructor is empty, it probably doesn't hurt not having one.

---

Can the `BoutMesh` constructor you've added just go in `FakeParallelMesh`? Or does it need access to `private` members of `BoutMesh`?

cmacmackin commented:
Yes, the constructor needs access to private members of `BoutMesh`.

cmacmackin commented:
I've split the MPI wrapper methods off of `BoutMesh` into their own class. As this involved some fairly substantial changes throughout the code base, I decided to submit these in a separate pull request which can be considered once this one has been merged. See PR #1808.

cmacmackin commented:
I've implemented a BoutReal conversion vector elements and confirmed that all tests compile and pass. I'd like to add something similar for matrix elements, but I'm not sure how best to approach it. Because these "elements" may in fact correspond to multiple elements weighted for interpolation, there isn't really a single value which can be returned. The options I can see are:
 
1. Return the single i, j value anyway
2. Throw an error if the object is a setter for more than one element
3. Instead of a single value, return a vector of BoutReals

-----------------------------------------------------

I'm also not sure that the default copy/move assignment operators for the element classes give the behaviour we'd want. We would want the code

vectorObj(i) = vectorObj(j) = 5.;

to set both element i and j to 5, but if the default assignment is used then it would just set the ith Element to point at the jth Element without changing its value. Perhaps the assignment operators should be deleted or explicitly implemented.

-------------------------------------------------------

I'd like to know the thoughts of @ZedThree on how to proceed on both of these issues.

ZedThree commented:
On the first point, is it the case that the weighted matrix is for a particular use case (the AMG)? Or is it more general? Perhaps a better question is, is a matrix that just stores single values useful?

If so, it might make sense to have a more specialised implementation for the weighted matrix that returns a vector (or a more explicit type), and a base class that returns a single number.

Does that make sense for your use case?

---

I suspect an explicit implementation of the copy assignment operator makes more sense than deleting it. Your code snippet looks reasonable to me (i.e. there's no reason to disallow it), and your interpretation of what it should do also seems correct.
The copy constructor should then be explicitly `= default`, which we can see from:

```cpp
PetscVector::Element element{vector(i)};
```

If we understand `PetscVector::Element` to be a non-owning view of an element, then `element` should just be another reference to `vector(i)`, which the default copy constructor would give us.
(assuming I'm understanding everything correctly!)

The move constructor/assignment should then be marked [`=delete`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-five), as I'm not sure that operation either makes sense or can be implemented differently to copy.

cmacmackin commented:
>On the first point, is it the case that the weighted matrix is for a particular use case (the AMG)? Or is it more general? Perhaps a better question is, is a matrix that just stores single values useful?
>
>If so, it might make sense to have a more specialised implementation for the weighted matrix that returns a vector (or a more explicit type), and a base class that returns a single number.
>
>Does that make sense for your use case?

The weighted matrix is required for any use-case where differentiation is performed in the field-aligned direction. While this would not always be the case, it is not unique to AMG. I'm not sure whether it would justify keeping this in a single class or the creation of a subclass.

Is it possible to override a method with one of a different return-type?

ZedThree commented:
> Is it possible to override a method with one of a different return-type?

Yes, this is fine. Whether or not it's the cleanest method, I'm still not sure. Probably not if there's places where it would make sense to use either a "plain" `PetscMatrix` or a "weighted" version.
It might do if that isn't the case.

An alternative might be to have `PetscMatrix::Element::operator BoutReal` return the single i, j value, and have another method like `PetscMatrix::Element::getWeights`? (or whatever the appropriate name is).

Can you point me at a snippet where the weights from an element are used?

cmacmackin commented:
With respect to the constructors/assignment operators, I agree with your suggestions. However, the move constructor/assignment does not need to be deleted, as the default implementations provides a more efficient way of moving around the information needed for the "non-owning view" of the element. In fact, I can't return an object of type `Element` from my `operator()(ind_type& index)` methods without the move constructor.

cmacmackin commented:
> Can you point me at a snippet where the weights from an element are used?

I think perhaps there might be some misunderstanding between us. The issue isn't that in some places the weights from the elements are used. Rather, that in some cases when you assign a `BoutReal` to an element what actually happens is it assigns that value to multiple elements, multiplied by different weightings. For example, this happens [here in Laplace3dAmg](https://github.com/boutproject/BOUT-dev/blob/115a6ac7308528a4597fdf17692aeb51f8dd5cb4/src/invert/laplace/impls/petsc3damg/petsc3damg.cxx#L444), although you can't really see that's what's going on. In effect what is happening in the assignment `operator3D.yup()(l, lyp) = C_df_dy[l] + C_d2f_dy2[l]` is
```C++
operator3D(l, lyp.xm()) = (C_df_dy[l] + C_d2f_dy2[l]) * weight1;
operator3D(l, lyp) = (C_df_dy[l] + C_d2f_dy2[l]) * weight2;
operator3D(l, lyp.xp()) = (C_df_dy[l] + C_d2f_dy2[l]) * weight3;
// ...
```
Note that the same matrix can have this sort of weighted assignment and conventional assignment to a single element, as is the case in the `Laplace3dAmg` class I linked to. Thus it probably does not make sense to have separate matrix classes, in my mind.

cmacmackin commented:
I've decided that the simplest and most sensible thing to do is just to return whatever value was assigned to the matrix element on the RHS of the assignment operator. I.e., after executing `mat(i, j) = 3.;` then `static_cast<BoutReal>(mat(i,j))` would return `3.`. I think this is probably the most sensible approach to take and certainly the simplest.

cmacmackin commented:
As I can't see anywhere else the new `BoutMesh` constructor can go, I'm going to leave it where it is.

cmacmackin commented:
I've been examining why the CI still failed after I fixed those compilation issues. After some experimentation on my own machine, I discovered that the tests where I deliberately check erroneous behaviour in the PETSc interface behave unpredictably when linking against a non-debug version of PETSc. This will cause some tests to fail because PETSc doesn't return a non-zero error code and thus the interface does not throw an exception. Worse, some tests will outright crash, ending the whole program, because MPI_Abort gets called.

Ideally these tests could be wrapped in `#ifdef`s checking whether a debug-version of PETSc has been used, but there aren't any macros like this defined. The simplest option would simply be to delete or comment them out.

ZedThree commented:
Looks like `PETSC_USE_DEBUG` is defined in `petscconf.h`, we could use that.

cmacmackin commented:
Ah, perfect!

On Tue, Oct 22, 2019 at 4:12 PM Peter Hill <notifications@github.com> wrote:

> Looks like PETSC_USE_DEBUG is defined in petscconf.h, we could use that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boutproject/BOUT-dev/pull/1805?email_source=notifications&email_token=AB6ESPIR6BK2C56D2VJKA2LQP4J37A5CNFSM4JANPNM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB6DIUQ#issuecomment-545010770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6ESPPCHV3V6HHFKLGEO43QP4J37ANCNFSM4JANPNMQ>
> .
>


-- 
   Chris MacMackin

cmacmackin commented:
I found that those CI builds which compiled with OpenMP were failing their tests because [PETSc is not thread-safe](https://www.mcs.anl.gov/petsc/miscellaneous/petscthreads.html) and because of one loop for my indexer classes which is inherently serial. I have updated the code to make a few loops always be serial and to place the offending PETSc function calls in critical blocks.

The fact that the CMake build was passing even though it did compile with OpenMP indicated that it was not compiling the PETSc interface or tests. I have amended the `CMakeLists.txt` files so that they will now be built.

There was also one build without PETSc. This failed because I had tried to include the `petscconf.h` header from outside the `#ifdef BOUT_HAS_PETSC` region. I moved that include statement within the `ifdef`. As I write this I am waiting on the CI server to complete, but am hopeful that it should now pass.

cmacmackin commented:
I've finally gotten the CI test to all pass, having eliminated the data races and other thread-unsafe behaviour within the PETSc interface and tests. I've run `clang-format` on the files I wrote and will push those changes shortly. However, before I do, I wondered whether I should apply `clang-format` to all files in the repository. However, before I do, I'd like to know whether there are any other changes which remain to be made before this branch can be merged into `next`.

d7919 commented:
I think generally we'd ask that you clang format just the regions that you've touched rather than whole files but @ZedThree can probably provide exact guidance.

ZedThree commented:
I wouldn't run `clang-format` over anything except the changes in this PR. There will be a _lot_ of other changes otherwise! I'm pretty sure there's a way to limit it to just particular commits (via `git clang-format`), but I'm happy to investigate that myself if it proves non-obvious.

I'm in the final stages of prepping the v4.3.0 release, so I hope to be able to give this another read through either tomorrow or Monday.

cmacmackin commented:
Thanks for the heads-up about `git clang-format`. I wasn't aware that existed but it made life very easy. I just ran `git clang-format --extension cxx,hxx next` (i.e., applying it to all changes in .cxx and .hxx files relative to the `next` branch). I'll push those changes now so you can take a look at them tomorrow or Monday.

ZedThree commented on include/bout/petsc_interface.hxx:
```diff
@@ -0,0 +1,600 @@
+  }
+
+  /// Returns a field constructed from the contents of this vector
+  T toField() {
```
Can mark this `const`

ZedThree commented on include/bout/petsc_interface.hxx:
```diff
@@ -0,0 +1,600 @@
+      }
+    };
+    MatCreate(comm, matrix.get());
+    MatSetSizes(*matrix, size(), size(), PETSC_DECIDE, PETSC_DECIDE);
```
See comment above about `size` in `PetscVector`

ZedThree commented on src/mesh/mesh.cxx:
```diff
@@ -317,6 +318,89 @@ bool Mesh::hasBndryUpperY() {
+int Mesh::globalStartIndex3D() {
+  int localSize = localSize3D();
+  int cumulativeSize;
+  MPI_Scan(&localSize, &cumulativeSize, 1, MPI_INT, MPI_SUM, BoutComm::get());
```
I guess these functions are not going to be called a lot? Otherwise, we probably want to cache the value.

cmacmackin replied:
They will only be called when creating a new indexer object for the mesh. In the case of the global mesh, this will only happen once (at startup). We could cache it for completeness, but in reality it won't be called again anyway.

ZedThree commented on src/sys/petsc_interface.cxx:
```diff
@@ -0,0 +1,173 @@
+Mesh* GlobalIndexer::getMesh() { return fieldmesh; }
+
+PetscInt GlobalIndexer::getGlobal(Ind2D ind) {
+  return static_cast<PetscInt>(indices2D[ind] + 0.5);
```
You can use `std::round` instead of the trick with `0.5`. The `static_cast` is still needed to get rid of a warning about implicit conversions

ZedThree commented on tests/unit/fake_parallel_mesh.hxx:
```diff
@@ -0,0 +1,288 @@
+      }
+      if (j == 0) {
+        meshes.at(j + i * nype).local3D += (nx - 2) * nz;
+        // meshes.at(j + i*nype).local2D += (nx - 2);
```
Commented out code, and at L271

cmacmackin commented on src/sys/petsc_interface.cxx:
```diff
@@ -0,0 +1,173 @@
+      indices3D(it.ind, localmesh->yend + 1, z) = counter++;
+    }
+  }
+  BOUT_FOR_SERIAL(i, localmesh->getRegion3D("RGN_NOY")) {
```
I'm thinking I might do this now, actually, as I find myself needing to iterate over boundary regions on a number of different occasions. The existing `RGN_INNER_X`, `RGN_LOWER_Y`, etc. almost do the job, except they include more than one cell in the x-direction and they don't include corner cells. It would be quite easy to create some additional regions for this purpose in `BoutMesh::addBoundaryRegions()`, although I'm not quite sure what to call them.

ZedThree replied:
I'm not sure we've got a good term for "the first point in the boundary". Maybe even just `RGN_LOWER_Y_FIRST_POINT_ONLY` as a placeholder for now?

I suspect we will need to come up with some good terminology soon, as those regions will also be needed for boundary conditions where we want each point in the plane of the boundary, plus some arbitrary number of points perpendicular to it.

Other suggestions:

- `RGN_*_PLANE`
- `RGN_*_SLICE`
- `RGN_*_PRIMARY`
- `RGN_*_HEAD`
- `RGN_*_FRONT`
- `RGN_*_MINOR`
- `RGN_*_BEGIN`

`begin` is quite appealing -- mirrors the names for iterators. Of course, it does imply `RGN_*_END` which is probably not as useful

cmacmackin commented on include/bout/petsc_interface.hxx:
```diff
@@ -0,0 +1,600 @@
+  static IndexerPtr globalInstance;
+  static bool initialisedGlobal;
+  static Mesh* globalmesh;
+  bool initialised;
```
Clang was right. I think I was going to use it for something and then realised I didn't need to.

cmacmackin commented on include/bout/petsc_interface.hxx:
```diff
@@ -0,0 +1,600 @@
+    BOUT_FOR(i, f.getRegion(RGN_ALL)) {
+      PetscInt ind = indexConverter->getGlobal(i);
+      if (ind != -1) {
+        BOUT_OMP(critical) VecSetValues(*vector, 1, &ind, &f[i], INSERT_VALUES);
```
Either would work. I was thinking that since there are a few non-critical statements in the loop parallelism might give some performance boost but on reflection probably not (given the overhead involved).

I wasn't aware of `VecSetValue` as it wasn't mentioned in the PDF PETSc manual. I've switched over to that where appropriate. Longer term, it may be possible to optimise this bit of code by using `VecSetValues` wherever there is a continuous stretch of field data which should be copied.

cmacmackin commented on src/sys/petsc_interface.cxx:
```diff
@@ -0,0 +1,173 @@
+      if (it.ind == localmesh->xstart)
+        indices3D(it.ind - 1, localmesh->ystart - 1, z) = counter++;
+      if (it.ind == localmesh->xend)
+        indices3D(it.ind + 1, localmesh->ystart - 1, z) = counter++;
```
In principle this doesn't matter but I agree it's messy. However, since I refactored to use new Region objects this is a moot point. The regions will have everything in the obvious order.

cmacmackin commented on include/bout/petsc_interface.hxx:
```diff
@@ -0,0 +1,600 @@
+
+  /// Construct a vector like v, but using data from a raw PETSc
+  /// Vec. That Vec (not a copy) will then be owned by the new object.
+  PetscVector(const PetscVector<T> v, Vec* vec) {
```
Yes, that was a typo/oversight on my part.

cmacmackin commented on include/bout/petsc_interface.hxx:
```diff
@@ -0,0 +1,600 @@
+ *  Performs matrix-multiplication on the supplied vector
+ */
+template <class T>
+PetscVector<T> operator*(PetscMatrix<T>& mat, PetscVector<T>& vec) {
```
I renamed the getter routines as you suggested and added `const` overloads for them so that the matrix multiplication operator can take `const` arguments.

cmacmackin commented:
I've committed the changes requested above, as well as a few bug-fixes I made while running integration tests for the 3D AMG Laplace solver. I've also merged in the latest changes from the `next` branch. Once the CI is finished running everything should be ready to merge.

cmacmackin commented:
2 of the CI runs failed, but I'm not sure why. In both cases the log says
```
$ travis_apt_get_update

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated

```
It looks like something may be wrong with Travis itself.

bendudson commented:
Yes, agreed. It looks like Travis had a hiccup. I've restarted the jobs so hopefully they work this time.

ZedThree commented:
For reason the submodules are reverting to the previous commit we had. But I can see `next` was merged in very recently and merging it in again doesn't update the submodules. I'm going to sort it out manually and push to this branch.

cmacmackin commented:
Fixed those. It seems I accidentally introduced a commit reverting the updates to the submodules. I had thought that I was bringing the submodules commits in my repo up-to-date with the ones in `next` but I actually overwrote those changes with the old commits by mistake.

ZedThree commented:
Great work, thanks @cmacmackin!
